src/Authentication/Authentication/custom/common/Permissions.ps1

```diff
@@ -1,36 +1,49 @@
 # ------------------------------------------------------------------------------
 #  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
 # ------------------------------------------------------------------------------
-
-
-$permissionsPulledFromMgGraphRequest = $null
+Update-FormatData -PrependPath $PSScriptRoot\..\..\Microsoft.Graph.Authentication.format.ps1xml
 
 function Permissions_GetPermissionsData {
+    param (
+        [bool] $online
+    )
 
     $permissions_MsGraphServicePrincipal = $null
+    $requestError = $null
+    $fromInvokeMgGraphRequest = $false
 
     # 2. Making a REST request to MS Graph
-    if ($null -eq $permissions_MsGraphServicePrincipal){
+
+    if (($null -eq $permissions_MsGraphServicePrincipal) -or ($null -ne $permissions_MsGraphServicePrincipal -and $fromInvokeMgGraphRequest -eq $false)) {
         $script:permissions_MsGraphServicePrincipal = try {
 
             # Write-Host "Getting data from web service"
             $result = Invoke-MgGraphRequest -method GET 'https://graph.microsoft.com/v1.0/servicePrincipals?filter=appId eq ''00000003-0000-0000-c000-000000000000''' 
             
             if ($null -ne $result) {
                 $result | select-object -expandproperty value 
+                $script:fromInvokeMgGraphRequest = $true
```

Comment:
Also, thi should be named with the `permissions_` prefix to show that this variable is part of the `Permissions` class.